src/paths/steiner.cpp

```diff
@@ -0,0 +1,251 @@
+#include <math.h>
+#include <map>
+#include <limits.h>
+#include <vector>
+#include <set>
+
+std::map<std::set<igraph_integer_t>, igraph_integer_t> subsetMap;
+
+void printSubsets(std::set<std::set<igraph_integer_t>> allSubsets)
+{
+	printf("Subsets :\n{\n");
+	for (auto i = allSubsets.begin() ; i != allSubsets.end() ; ++i )
+	{
+		printf("\t{ ");
+		for (auto j = (*i).begin() ; j != (*i).end() ; j++)
+		{
+			if (j != (*i).begin()) { printf(", ");}
+			printf("%ld",*j);
+		}
+		printf("}");
+		if (i != (--allSubsets.end())) { printf(",\n");}
+	}
+	printf("\n}\n");
+}
+std::set<std::set<igraph_integer_t>> generateSubsets(igraph_vector_int_t steinerTerminals, igraph_integer_t n, igraph_integer_t graphsize)
+{
+	igraph_integer_t count = (1 << n);
+	std::set<std::set<igraph_integer_t>> allSubsets;
+	igraph_integer_t subsetIndex = graphsize;
+
+	// The outer for loop will run 2^n times to print all subset .
+	// Here variable i will act as a binary counter
+
+	for (igraph_integer_t i = 0; i < count; i++)
+	{
+		// The inner for loop will run n times , As the maximum number of elements a set can have is n
+		// This loop will generate a subset
+		std::set<igraph_integer_t> newSubset;
+		for (igraph_integer_t j = 0; j < n; j++)
+		{
+			// This if condition will check if jth bit in binary representation of  i  is set or not
+			// if the value of (i & (1 << j)) is greater than 0 , include arr[j] in the current subset
+			// otherwise exclude arr[j]
+			if ((i & (1 << j)) > 0)
+			{
+				newSubset.insert(VECTOR(steinerTerminals)[j]);
+			}
+		}
+		
+		if (newSubset.size() > 1)  
+		{
+			if (allSubsets.find(newSubset) == allSubsets.end())
+			{
+				allSubsets.insert(newSubset);
+				subsetMap.insert(std::make_pair(newSubset, subsetIndex));
+				subsetIndex++;
+			}
+		}
+	}
+	return allSubsets;
+}
+
+igraph_integer_t fetchIndexofMapofSets(std::set<igraph_integer_t> subset)
+{
+	igraph_integer_t key;
+	std::map<std::set<igraph_integer_t>, igraph_integer_t>::iterator it;
+	for (it = subsetMap.begin(); it != subsetMap.end(); ++it)
+	{
+		if (it->first == subset)
+		{
+			key = it->second;
+		}
+	}
+
+	return key;
+}
+
+igraph_error_t igraph_steiner_dreyfus_wagner(const igraph_t *graph,const igraph_vector_int_t* steiner_terminals,
+igraph_neimode_t mode, const igraph_vector_t *weights)
+{
+
+	igraph_integer_t no_of_vertices = (igraph_integer_t)igraph_vcount(graph);
+	igraph_integer_t no_of_edges = igraph_ecount(graph);
+
+	igraph_vector_t steiner_vertices;
+	igraph_vector_int_t steiner_terminals_copy; 
+	igraph_matrix_t dp_cache; // dynamic programming table
+	igraph_integer_t q;
+	std::set<std::set<igraph_integer_t>> allSubsets;
+	igraph_matrix_t distance;
+	IGRAPH_VECTOR_INIT_FINALLY(&steiner_vertices, 0);
+
+	if (igraph_vector_size(weights) != no_of_edges)
+	{
+		char error[1024] = {0};
+		sprintf(error,"Weight vector length does not match %ld vec size and %ld edges \n",igraph_vector_size(weights),no_of_edges);
+		IGRAPH_ERROR(error, IGRAPH_EINVAL);
```

Comment:
Use `IGRAPH_ERRORF` instead of `sprintf` + `IGRAPH_ERROR`, and make sure to use `"%" IGRAPH_PRId` instead of `"%ld"` when printing an `igraph_integer_t`.

Reply:
Hi, I tried using `"%" IGRAPH_PRId` but it doesn't recognize it. Will you be able to show an example of how it's used? I couldn't find it in the develop section Reference Manual.

Reply:
`IGRAPH_PRId` is defined here: https://github.com/igraph/igraph/blob/develop/include/igraph_types.h#L64

There are plenty of usage examples for `IGRAPH_PRId` in the codebase, for instance, here:

https://github.com/igraph/igraph/blob/develop/tests/unit/igraph_rng_get_integer.c#L33

You made a good point with the documentation, though - I think it's not mentioned anywhere in the docs yet, only in the examples, so we should fix that.